### ISSUE TITLE
Calls (like TwingleDonation.cancel) fails if contribution is not found

### ISSUE THREAD

**Issue by bjendres** [systopia/de.systopia.twingle#30]
If the contribution with the given Twingle-ID is not found, the API returns an error. That might sound like the proper response, but this behaviour causes Twingle to submit the same request over and over.

There might be a point in flagging this issue to a user (e.g. by E-Mail) and returning an "ok" in the API.

**Comment by jensschuppe** [comment]
To me that seems to be an error-management issue for Twingle, not necessarily for the CiviCRM API, since how to handle that will highly depend on what your workflows are: activities, e-mails, ignore, ...

Also, reporting by e-mail does not stop Twingle to re-send it again and again. Twingle users should really be checking those error messages regularyl. Maybe this could be outlined in the API docs.

**Comment by bjendres** [comment]
> To me that seems to be an error-management issue for Twingle, not necessarily for the CiviCRM API, since how to handle that will highly depend on what your workflows are: activities, e-mails, ignore, ...

I agree, but that's the way is currently implemented: if there is an error: re-submit with the next batch.

> Also, reporting by e-mail does not stop Twingle to re-send it again and again. 

Yes ist does, if you don't return an error any more.

**Comment by jensschuppe** [comment]
> > Also, reporting by e-mail does not stop Twingle to re-send it again and again.

> Yes ist does, if you don't return an error any more.

Agreed. What happens in CiviCRM should be configurable then, because not throwing an error will make it vanish from CiviCRM logs as well. I see the following options:

- Ignore
- Log to the CiviCRM log
- Schedule an activity
- Send an e-mail
- maybe plug into other systems (trigger a CiviRules event, execute an action-provider action, etc.), maybe provide a hook.

**Comment by bjendres** [comment]
Will be released with ``1.2``

**Comment by webworka** [comment]
For Information from the twingle side: We released last week an update on the interface that we ignore this error on our side.

Normally the handling is that if an error is given back vom CiviCRM we will send this request again. In this case, this makes no sense. 
I also found another error which we now also ignore: "SEPA Mandate xxxx already terminated". From the twingle point of view this is the same type of error, that should not cause the interface to stop.

**Comment by jensschuppe** [comment]
@webworka Thanks for reporting back here, too!

I think it would be reasonable, though, to _clarify the error messages as a first step_, because ignoring errors is not a good thing in general, and those come from the CiviCRM core API and are not necessarily bound to only those two events. Ultimately, we should not issue errors back to you (Twingle) in that case at all, but handle such inconsistencies within CiviCRM.

We will be implementing the _e-mail thing as a second step_, so that admins can be notified and Twingle does not need to care about that anymore.
